File: 2023/day01/day01.py
import string
from functools import reduce
from typing import Optional

def calibration_value(line: str) -> int:
    digits = [d for d in line if d in string.digits]
    return int(digits[0] + digits[-1])

def digit_word(s: str, index: int) -> Optional[int]:
    lookup = {
        'one': 1,
        'two': 2,
        'three': 3,
        'four': 4,
        'five': 5,
        'six': 6,
        'seven': 7,
        'eight': 8,
        'nine': 9,
    }
    for name, value in lookup.items():
        if s[index:index + len(name)] == name:
            return value
    return None

def advanced_calibration_value(line: str) -> int:
    digits = []
    index = 0
    while index < len(line):
        if line[index] in string.digits:
            digits.append(line[index])
            index += 1
            continue
        possible_spelling = digit_word(line, index)
        if possible_spelling:
            digits.append(str(possible_spelling))
        index += 1
    return int(digits[0] + digits[-1])


def product_of_calibration(lines: list[str]) -> int:
    values = [calibration_value(line) for line in lines]
    return reduce(lambda acc, new: acc * new, values, 1)

def sum_of_calibration(lines: list[str], func=calibration_value) -> int:
    values = [func(line) for line in lines]
    return reduce(lambda acc, new: acc + new, values, 0)

def part1(filename):
    with open(filename) as f:
        return sum_of_calibration(lines=f.readlines(), func=calibration_value)

def part2(filename):
    with open(filename) as f:
        return sum_of_calibration(lines=f.readlines(), func=advanced_calibration_value)

    
if __name__=='__main__':
    print(part1('input.txt'))
    print(part2('input.txt'))

File: 2023/day01/day01.hs
module Day01 where

import Data.Char
import Data.Map (Map)
import qualified Data.Map as M

digitNames = [
    ("one", 1),
    ("two", 2),
    ("three", 3),
    ("four", 4),
    ("five", 5),
    ("six", 6),
    ("seven", 7),
    ("eight", 8),
    ("nine", 9)
    ]


getDigits :: String -> [Int]
getDigits "" = []
getDigits (c:cs)
  | elem c "123456789" = (ord c - ord '0'):(getDigits cs)
  | otherwise =  getDigits cs 

digitName :: String -> Maybe (Int, String)
digitName s 
  | possible == [] = Nothing
  | otherwise = Just (value, drop (length name) s)
  where possible = take 1 $ filter (\(n,v) -> n == (take (length n) s)) digitNames
        (name, value) = head possible


getDigitsAdvanced :: String -> [Int]
getDigitsAdvanced "" = []
getDigitsAdvanced (d:ds)
  | elem d "123456789" = (ord d - ord '0'):(getDigitsAdvanced ds)
  | otherwise = case (digitName (d:ds)) of
                    Just (value, rem) -> value:(getDigitsAdvanced ds)
                    Nothing -> getDigitsAdvanced ds

calibrationValue :: String -> Int
calibrationValue s = (head nums) * 10 + (last nums)
  where nums = getDigits s


advancedCalibrationValue :: String -> Int 
advancedCalibrationValue s = (head nums) * 10 + (last nums)
  where nums = getDigitsAdvanced s 


sumCalibrations :: (String -> Int) -> [String] -> Int
sumCalibrations func lines = sum $ map func lines


part1 :: String -> IO ()
part1 filename = do 
                    contents <- lines <$> readFile filename
                    let answer = sumCalibrations calibrationValue contents
                    print(answer)


part2 :: String -> IO ()
part2 filename = do 
                    contents <- lines <$> readFile filename
                    let answer = sumCalibrations advancedCalibrationValue contents
                    print(answer)